src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues, columnNulls, NULL);
+
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			/* find the partition column value */
+
+			if (columnNulls[partitionColumn->varattno - 1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+
+			partitionColumnValue = columnValues[partitionColumn->varattno - 1];
+
+			/* find the shard interval and id for the partition column value */
+			shardInterval = FindShardInterval(partitionColumnValue, shardIntervalCache,
+											  shardCount, partitionMethod,
+											  compareFunction, hashFunction,
+											  useBinarySearch);
+			if (shardInterval == NULL)
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("could not find shard for partition column "
+									   "value")));
+			}
+
+			shardId = shardInterval->shardId;
+
+			MemoryContextSwitchTo(oldContext);
+
+			/* find the connections to the shard placements */
+			shardConnections = GetOrOpenShardConnections(shardConnectionHash,
+														 shardId,
+														 copyStatement,
+														 &shardConnectionsFound);
+			if (!shardConnectionsFound)
+			{
+				/* send binary headers to shard placements */
+				resetStringInfo(copyOutState->fe_msgbuf);
+				AppendCopyBinaryHeaders(copyOutState);
+				SendCopyDataToAll(copyOutState->fe_msgbuf,
+								  shardConnections->connectionList);
+			}
+
+			/* replicate row to shard placements */
+			resetStringInfo(copyOutState->fe_msgbuf);
+			AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+							  copyOutState, columnOutputFunctions);
+			SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+
+			processedRowCount += 1;
+		}
+
+		connectionList = ConnectionList(shardConnectionHash);
+
+		/* send binary footers to all shard placements */
+		resetStringInfo(copyOutState->fe_msgbuf);
+		AppendCopyBinaryFooters(copyOutState);
+		SendCopyDataToAll(copyOutState->fe_msgbuf, connectionList);
+
+		/* close the COPY input on all shard placements */
+		EndRemoteCopy(connectionList, true);
+
+		if (CopyTransactionManager == TRANSACTION_MANAGER_2PC)
+		{
+			PrepareRemoteTransactions(connectionList);
+		}
+
+		EndCopyFrom(copyState);
+		heap_close(distributedRelation, NoLock);
+
+		error_context_stack = errorCallback.previous;
+
+		/* check for cancellation one last time before committing */
+		CHECK_FOR_INTERRUPTS();
+	}
+	PG_CATCH();
+	{
+		List *abortConnectionList = NIL;
+
+		error_context_stack = errorCallback.previous;
+
+		/* roll back all transactions */
+		abortConnectionList = ConnectionList(shardConnectionHash);
+		EndRemoteCopy(abortConnectionList, false);
+		AbortRemoteTransactions(abortConnectionList);
+		CloseConnections(abortConnectionList);
```

Comment:
I was thinking the connections are not really remote in the way that transactions are remote.